### PR TITLE
mmds: added storing data in MMDS at startup

### PR DESCRIPTION
**Pull request by georgepisaltu**
Signed-off-by: George Pisaltu <gpl@amazon.com>

## Reason for This PR

#1292 

## Description of Changes

Firecracker can now start with data in MMDS from a JSON file provided in command line arguments.

Currently just a draft, needs integration and unit testing.

Edit: added integration testing for this feature and converted from draft to open PR.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

### CODE REVIEW

**Review by serban300 — APPROVED** [review]

**Comment by raduiliescu** [comment]
While the change is ok, it doesn't have a test

File: tests/integration_tests/functional/test_mmds.py
Comment (sandreim) [inline tests/integration_tests/functional/test_mmds.py:364]: I would also add a negative unit test like for example if the metadata file is missing or bad permissions.

File: tests/integration_tests/functional/test_mmds.py
Comment (sandreim) [inline tests/integration_tests/functional/test_mmds.py:20]: Does it make sense to add this functionality in framework/microvm.py ?

**Review by sandreim — COMMENTED** [review]
Added a few comments. Looks good. Needs a rebase.

**Review by acatangiu — APPROVED** [review]

File: tests/integration_tests/functional/test_mmds.py
Comment (gc-plp) [reply tests/integration_tests/functional/test_mmds.py:20]: It does, yes. It's a microVM related property and it could be used in other tests too.

File: tests/integration_tests/functional/test_mmds.py
Comment (gc-plp) [inline tests/integration_tests/functional/test_mmds.py:26]: Is this block simply copying a file to a path? We can do that in one line using `shutil.copy()`.

**Comment by dianpopa** [comment]
Currently blocked on #2238.

**Comment by richardcase** [comment]
Shame this was closed as its exactly what we are after.

**Comment by georgepisaltu** [comment]
@richardcase from our talk on Slack you seemed to be willing to contribute here. We decided to unblock this PR and it would be great if you want to pick this up.

It would be a great enhancement to our metadata service! Please let us know what your plans are here.

**Comment by richardcase** [comment]
Thanks @georgepisaltu . Yes i'd be happy to contribute here. I will have a think over the weekend and come up with a plan. Shall i comment here or on the issue?

**Comment by georgepisaltu** [comment]
@richardcase awesome!

Let's use issue #1292 for further discussion.

**Pull request closed without merging**
